Libraries/LibGUI/FilteringProxyModel.h

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright (c) 2020, the SerenityOS developers.
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this
+ *    list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+
+#include <AK/NonnullRefPtr.h>
+#include <AK/Optional.h>
+#include <AK/String.h>
+#include <LibGUI/Model.h>
+#include <LibGUI/TextBox.h>
+
+namespace GUI {
+
+class FilteringProxyModel final : public Model {
+public:
+    static NonnullRefPtr<FilteringProxyModel> construct(Model& model)
+    {
+        return adopt(*new FilteringProxyModel(model));
+    }
+
+    virtual ~FilteringProxyModel() override {};
+
+    virtual int row_count(const ModelIndex& = ModelIndex()) const override;
+    virtual int column_count(const ModelIndex& = ModelIndex()) const override;
+    virtual Variant data(const ModelIndex&, Role = Role::Display) const override;
+    virtual void update() override;
+    virtual ModelIndex index(int row, int column = 0, const ModelIndex& parent = ModelIndex()) const override;
+
+    void start_filtering() { m_is_filtering = true; }
+    void stop_filtering();
+    void set_filter_term(const String& term) { m_filter_term = term; }
+
+    bool is_filtering() const { return m_is_filtering; }
+
+    ModelIndex map(const ModelIndex&) const;
+
+private:
+    void filter();
+    FilteringProxyModel(Model& model)
+        : m_model(model)
+    {
+    }
+
+    Model& m_model;
+
+    // Maps row to actual model index.
+    Vector<ModelIndex> m_matching_indices;
+
+    bool m_is_filtering { false };
```

Comment:
There should be no need for this member; we can just use `m_filter_term`'s null state.